src/Test/StateMachine/Internal/Utils/BoxDrawer.hs

```diff
@@ -0,0 +1,70 @@
+module Test.StateMachine.Internal.Utils.BoxDrawer
+  ( EventType(..)
+  , exec)
+  where
+
+import           Text.PrettyPrint.ANSI.Leijen
+                   (Doc, text, vsep)
+
+type Pid = Int
+
+data EventType = Open | Close
+  deriving (Show)
+
+data Event = Event EventType Pid String
+
+data Cmd = Top | Start String | Active | Deactive | Ret String | Bottom
+
+compile :: [Event] -> ([Cmd], [Cmd])
+compile = go (Deactive, Deactive)
+  where
+    infixr 9 `add`
+    add :: (a,b) -> ([a], [b]) -> ([a], [b])
+    add (x,y) (xs, ys) = (x:xs, y:ys)
+
+    set :: (a, a) -> Pid -> a -> (a, a)
+    set (_x, y) 1 x' = (x', y)
+    set (x, _y) 2 y' = (x, y')
+    set _ pid _ = error $ "compile.set: unknown pid " ++ show pid
+
+    go :: (Cmd, Cmd) -> [Event] -> ([Cmd], [Cmd])
+    go _ [] = ([], [])
+    go st (Event Open pid l :  rest) =
+      set st pid Top `add` set st pid (Start l) `add` go (set st pid Active) rest
+    go st (Event Close pid l :  rest) =
+      set st pid (Ret l) `add` set st pid Bottom `add` go (set st pid Deactive) rest
+
+size :: Cmd -> Int
+size Top       = 4
+size (Start l) = 5 + length l
+size Active    = 2
+size Deactive  = 0
+size (Ret l)   = 10 + length l
+size Bottom    = 4
+
+adjust :: Int -> Cmd -> String
+adjust n Top = " ┌" ++ replicate (n - 4) '─' ++ "┐"
+adjust n (Start l) = " │" ++ l ++ replicate (n - length l - 5) ' ' ++ " │"
+adjust n Active = " │" ++ replicate (n - 4) ' ' ++ "│"
+adjust n Deactive = replicate (n - 1) ' '
+adjust n (Ret l) = " │" ++ replicate (n - 8 - length l) ' ' ++ "--> " ++ l ++ "│"
```

Comment:
We should perhaps also open a ticket on making the utf8 output optional (or even better, detect if utf8 is supported by the terminal, if not use ascii)?